FV3/io/FV3GFS_io.F90

```diff
@@ -2932,6 +2949,106 @@ end subroutine fv3gfs_diag_register_coarse
+    real(kind=kind_phys),      intent(in) :: delp(isco:ieco,jsco:jeco,1:levo)
+    character(len=64),         intent(in) :: coarsening_strategy
+    real(kind=kind_phys),      intent(in) :: ptop
+
+    logical :: require_area, require_masked_area, require_mass, require_masked_mass, require_vertical_remapping
+    real(kind=kind_phys), allocatable :: area(:,:)
+    real(kind=kind_phys), allocatable :: mass(:,:,:), phalf(:,:,:), phalf_coarse_on_fine(:,:,:)
+    real(kind=kind_phys), allocatable :: masked_area(:,:,:)
+
+    real(kind=kind_phys) :: var2d(nx, ny)
+    real(kind=kind_phys) :: var3d(nx, ny, levs)
+    integer :: i, j, ii, jj, k, isc, jsc, ix, nb, index, used
+
+    isc   = atm_block%isc
+    jsc   = atm_block%jsc
+
+    if (write_coarse_diagnostics) then
+      call determine_required_coarse_graining_weights(diag_coarse, coarsening_strategy, require_area, require_masked_area, require_mass, require_vertical_remapping)
+      if (.not. require_vertical_remapping) then
+        if (require_area) then
+          allocate(area(nx, ny))
+          call get_area(Atm_block, IPD_Data, nx, ny, area)
+        endif
+        if (require_mass) then
+          allocate(mass(nx, ny, levs))
+          call get_mass(Atm_block, IPD_Data, delp, nx, ny, levs, mass)
+        endif
+      else
+        allocate(area(nx, ny))
+        allocate(phalf(nx, ny, levs + 1))
+        allocate(phalf_coarse_on_fine(nx, ny, levs + 1))
+        allocate(masked_area(nx, ny, levs))
+        call get_area(Atm_block, IPD_Data, nx, ny, area)
+        call vertical_remapping_requirements(delp, area, ptop, phalf, phalf_coarse_on_fine)
+        call mask_area_weights(area, phalf, phalf_coarse_on_fine, masked_area)
+      endif
+   endif
+
+    do index = 1, DIAG_SIZE
+      if (trim(Diag(index)%name) .eq. '') exit
+      if (Diag(index)%id .gt. 0 .or. Diag_coarse(index)%id .gt. 0) then
+        if (Diag(index)%axes .eq. 2) then
+          do j = 1, ny
+            jj = j + jsc - 1
+            do i = 1, nx
+                ii = i + isc - 1
+                nb = Atm_block%blkno(ii,jj)
+                ix = Atm_block%ixp(ii,jj)
+                var2d(i,j) = Diag(index)%data(nb)%var2(ix)
+            enddo
+          enddo
+          if (Diag(index)%id > 0) then
+            used = send_data(Diag(index)%id, var2d, Time)
+          endif
+          if (Diag_coarse(index)%id > 0) then
+            call store_data2D_coarse(Diag_coarse(index)%id, Diag_coarse(index)%name, &
+              Diag_coarse(index)%coarse_graining_method, nx, ny, var2d, area, Time)
+          endif
+        elseif (Diag(index)%axes .eq. 3) then
+          do k=1, levs
+            do j = 1, ny
+              jj = j + jsc - 1
+              do i = 1, nx
+                  ii = i + isc - 1
+                  nb = Atm_block%blkno(ii,jj)
+                  ix = Atm_block%ixp(ii,jj)
+                  var3d(i,j,k) = Diag(index)%data(nb)%var3(ix,levs - k + 1)
```

Comment:
do physics and dycore always use opposite convention for vertical dimension order? or is there a flag for this?

Reply:
The dycore is written in a way that it can handle either through the `flip_vc` argument to various dynamics-physics coupling subroutines, but this physics suite always uses the convention that the lowest model level is at the bottom (it's hard-coded in `atmos_model.F90`):
https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/db45725a448eba0e305a0d34b4e5909dce06a8e9/FV3/atmos_model.F90#L230